internal/measurexlite/resolver.go

```diff
@@ -0,0 +1,106 @@
+package measurexlite
+
+//
+// Trusted Recursive Resolver
+//
+
+import (
+	"context"
+	"net"
+
+	"github.com/ooni/probe-cli/v3/internal/model"
+	"github.com/ooni/probe-cli/v3/internal/netxlite"
+)
+
+// NewTrustedRecursiveResolver2 returns a new trusted recursive resolver
+func NewTrustedRecursiveResolver2(logger model.Logger, URL string) model.Resolver {
+	const mozillaURL = "https://mozilla.cloudflare-dns.com/dns-query"
+	if URL == "" {
+		URL = mozillaURL
+	}
+	return &TrustedRecursiveResolver2{
+		Logger:                            logger,
+		URL:                               URL,
+		NewParallelDNSOverHTTPSResolverFn: nil,
+		ResolverSystem:                    netxlite.NewStdlibResolver(logger),
+	}
+}
+
+// TrustedRecursiveResolver2 emulates Firefox's TRR2 mode
+type TrustedRecursiveResolver2 struct {
+	Logger model.Logger
+	URL    string
+
+	// NewParallelDNSOverHTTPSResolverFn is OPTIONAL and can be used to overide
+	// calls to the netxlite.NewParallelDNSOverHTTPSResolver factory
+	NewParallelDNSOverHTTPSResolverFn func(model.Logger, string) model.Resolver
+
+	// ResolverSystem is MANADATORY and is used as a fallback if the DoH resolver fails
+	ResolverSystem model.Resolver
+}
+
+var _ model.Resolver = &TrustedRecursiveResolver2{}
+
+// Address implements model.Resolver.Address
+// we can't return a decisive address here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Address() string {
+	return ""
+}
+
+// Network implements model.Resolver.Network
+// we can't return a decisive network here since we don't know if LookupHost
+// succeeds with the DoH resolver or the system resolver
+func (trr *TrustedRecursiveResolver2) Network() string {
+	return ""
+}
+
+// CloseIdleConnections implements model.Resolver.CloseIdleConnections
+// we only check the fallback resolver since DoH resolver is generated on the fly
+func (trr TrustedRecursiveResolver2) CloseIdleConnections() {
+	trr.ResolverSystem.CloseIdleConnections()
+}
+
+// LookupHost implements model.Resolver.LookupHost
+func (trr *TrustedRecursiveResolver2) LookupHost(ctx context.Context, hostname string) ([]string, error) {
+	reso := trr.NewParallelDNSOverHTTPSResolverFn(trr.Logger, trr.URL)
+	if reso == nil {
+		reso = netxlite.NewParallelDNSOverHTTPSResolver(trr.Logger, trr.URL)
+	}
+	addrs, err := reso.LookupHost(ctx, hostname)
```

Comment:
I forgot to mention: if we want to implement TRR2 correctly, this lookup should be strictly time bounded (i.e., we should timeout the lookup after N seconds have elapsed)

Reply:
I have added a timeout to the DoH resolver here: https://github.com/ooni/probe-cli/pull/846/commits/286c78f1b0a2fadabd2119f89180cea43d8c8476#diff-4c58fee97f07a0f47542cb08705a81d5f43f76be1d6235538ceb3e4c51725e72R33,
this is currently configured to use Firefox's default timeout value of 1.5s